Sources/Core/Utilities/InAppPurchase/InAppPurchase.swift

```diff
@@ -0,0 +1,197 @@
+//
+//  InAppPurchase.swift
+//  PovioKit
+//
+//  Created by Marko Mijatovic on 20/01/2023.
+//  Copyright © 2023 Povio Inc. All rights reserved.
+//
+
+import Foundation
+import StoreKit
+
+
+@available(iOS 15.0, *)
+public final class InAppPurchase: NSObject {
+  public typealias IAPPlan = String
+  public typealias IAPReceipt = String
+  
+  private let productIdentifiers: [IAPPlan]
+  private var updateListenerTask: Task<Void, Error>? = nil
+  private var availableProducts: [Product] = []
+  private var purchasedProducts: [Product] = []
+  
+  /// Initialize new InAppPurchase with all available plans
+  /// - Parameter identifiers: Array of ``IAPPlan`` (eg. ["com.test.plan1", "com.test.plan2"])
+  public init(identifiers: [IAPPlan]) {
+    self.productIdentifiers = identifiers
+    super.init()
+    updateListenerTask = listenForTransactions()
+    
+    Task {
+      await requestProducts()
+      await updatePurchasedProducts()
+    }
+  }
+  
+  deinit {
+    updateListenerTask?.cancel()
+  }
+}
+
+// MARK: - Public
+@available(iOS 15.0, *)
+extension InAppPurchase {
+  /// Purchase plan with options
+  /// - Parameters:
+  ///   - plan: InAppPurchase plan identifier (eg. "com.test.plan") to purchase
+  ///   - options: Set of ``PurchaseOption``
+  /// - Returns: Result type with completed ``Transaction`` object or ``InAppPurchaseError``
+  /// - Important: ``InAppPurchaseError`` cases that this function returns:
+  /// * ``InAppPurchaseError.missingProductId``
+  /// * ``InAppPurchaseError.paymentCancelled``
+  /// * ``InAppPurchaseError.paymentPending``
+  /// * ``InAppPurchaseError.requestFailed(error)``
+  public func purchase(plan: IAPPlan, options: Set<Product.PurchaseOption> = []) async -> Result<Transaction, InAppPurchaseError> {
```

Comment:
Did you consider making the function throwable, instead of returning Result? What are pros/cons of either approach?

Reply:
Yes, I did, but I just like how the Result is handled in the implementation code, and no need for a do/catch block. It is cleaner for the user in my opinion. Do you have any thoughts on this?

Reply:
From what I've seen, the common approach is to use `async throws` in Apple's APIs and elsewhere. The advantage is, probably, that you don't have to handle error cases for every async expression; `do/catch` does it for you. But I don't have a strong preference at this point. 🤷